Share container lifetime implementation across all container tests

0xced opened the pull request:
## What does this PR do?

This PR makes all container test classes inherit from a new abstract `ContainerTest<TBuilderEntity, TContainerEntity>` class that takes care of implementing the `IAsyncLifetime` interface for starting and disposing the container. This reduces boilerplate code in all test projects.

## Why is it important?

This removes code duplication, thus reducing maintenance. Also, adding a timeout when starting a container during the tests could now be performed at a single place (`ContainerTest.cs`) instead of at 24 different places.

## Related issues

PR #908 is somehow related in that it simplifies the maintenance of the tests but both PRs can be merged independently.

0xced committed:
Share container lifetime implementation across all container tests

0xced commented:
I don't understand what's happening with the **Run Tests** step timing out after 6 hours. I downloaded the logs and parsed them with this quick and dirty code.

```cs
var logFile = new FileInfo(@"logs_2950\build (ubuntu-22.04)\8_Run Tests.txt"); 
using var logStream = logFile.OpenRead(); 
using var logReader = new StreamReader(logStream); 
var results = new List<TestResult>(); 
while (logReader.ReadLine() is {} line) 
{ 
    if (line.Contains("Passed:")) 
    { 
        var status = line.Substring(29, line.IndexOf('!') - 29); 
        var durationStart = line.IndexOf("Duration: ") + 10; 
        var duration = line.Substring(durationStart, line.LastIndexOf(" - ") - durationStart); 
        var assembly = line.Substring(line.IndexOf("net6.0/") + 7); 
        results.Add(new TestResult(status, duration, assembly.Substring(0, assembly.Length - 9))); 
    } 
} 

results.OrderBy(e => e.Assembly).Dump(); 

public record TestResult(string Status, string Duration, string Assembly);
```

And here are the results.

| Status  | Duration | Assembly                                  |
| :------ | :------- | :---------------------------------------- |
| Passed  | 233 ms   | Testcontainers.Azurite.Tests.dll          |
| Skipped | < 1 ms   | Testcontainers.CosmosDb.Tests.dll         |
| Passed  | < 1 ms   | Testcontainers.Couchbase.Tests.dll        |
| Passed  | < 1 ms   | Testcontainers.CouchDb.Tests.dll          |
| Passed  | 3 s      | Testcontainers.DynamoDb.Tests.dll         |
| Passed  | < 1 ms   | Testcontainers.Elasticsearch.Tests.dll    |
| Passed  | < 1 ms   | Testcontainers.EventStoreDb.Tests.dll     |
| Passed  | < 1 ms   | Testcontainers.K3s.Tests.dll              |
| Passed  | < 1 ms   | Testcontainers.Kafka.Tests.dll            |
| Passed  | 21 s     | Testcontainers.Keycloak.Tests.dll         |
| Passed  | 25 s     | Testcontainers.LocalStack.Tests.dll       |
| Passed  | 5 s      | Testcontainers.MariaDb.Tests.dll          |
| Passed  | 1 s      | Testcontainers.Minio.Tests.dll            |
| Passed  | 8 s      | Testcontainers.MongoDb.Tests.dll          |
| Passed  | 5 s      | Testcontainers.MsSql.Tests.dll            |
| Passed  | 12 s     | Testcontainers.MySql.Tests.dll            |
| Passed  | < 1 ms   | Testcontainers.Neo4j.Tests.dll            |
| Passed  | 30 s     | Testcontainers.Oracle.Tests.dll           |
| Passed  | 2 s      | Testcontainers.Platform.Linux.Tests.dll   |
| Skipped | 11 ms    | Testcontainers.Platform.Windows.Tests.dll |
| Passed  | 1 s      | Testcontainers.PostgreSql.Tests.dll       |
| Passed  | < 1 ms   | Testcontainers.RabbitMq.Tests.dll         |
| Passed  | < 1 ms   | Testcontainers.RavenDb.Tests.dll          |
| Passed  | 703 ms   | Testcontainers.Redis.Tests.dll            |
| Passed  | < 1 ms   | Testcontainers.Redpanda.Tests.dll         |
| Passed  | 647 ms   | Testcontainers.ResourceReaper.Tests.dll   |
| Passed  | < 1 ms   | Testcontainers.SqlEdge.Tests.dll          |
| Passed  | 50 s     | Testcontainers.Tests.dll                  |
| Passed  | 31 s     | Testcontainers.WebDriver.Tests.dll        |

There are 29 assemblies that were successfully tested. The one missing (`Testcontainers.Commons.dll`) targets .NET Standard and doesn't contain any tests. The logs look fine:

> No suitable test runtime provider was found:
>
> Skipping source: /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/tests/Testcontainers.Commons/bin/Debug/netstandard2.0/Testcontainers.Commons.dll (.NETStandard,Version=v2.0, X64)

The tests passed successfully in #908 so maybe they will pass if #908 gets merged first?

HofmeisterAn commented:
> I don't understand what's happening with the Run Tests step timing out after 6 hours.

I have not looked into the PR yet, I will review both PRs later today, but typically, this situation occurs when a wait strategy continuously fails. Currently, there is no default timeout or maximum number of retries in place. This is something I am currently working on. After looking into the logs this seams not the case for this pr 🤔. I need to take a closer look into it.

HofmeisterAn reviewed (commented):
I am not entirely sure what the root cause of the issue is, but the changes in `Testcontainers.Commons` are causing it:

> The one missing (`Testcontainers.Commons.dll`) targets .NET Standard and doesn't contain any tests. 

Although the logs indicate that no tests are being executed, the process actually hangs. At least, that is what happens on my machine when I run `dotnet test` within `tests\Testcontainers.Commons` directory. When I change the target framework to `net6.0` it works fine.

0xced committed:
Fix running tests hanging indefinitely

By setting the `IsTestProject` property to false in Testcontainers.Commons.csproj

Cause: referencing the `xunit` package enables `IsTestProject=true` which trigger a bug in vstest where [netstandard libraries hangs indefinitely][1]. This bug was addressed 5 days ago: [Fix execution gets stucks on single netstandard source][2]

[1]: https://github.com/microsoft/vstest/issues/4392
[2]: https://github.com/microsoft/vstest/pull/4497

0xced commented:
Thanks for your input. I now understand what's happening:

Referencing the `xunit` package enables `IsTestProject=true` which triggers a bug in vstest where [netstandard libraries hang indefinitely][1]. This bug was addressed only 5 days ago: [Fix execution gets stucks on single netstandard source][2]

[1]: https://github.com/microsoft/vstest/issues/4392
[2]: https://github.com/microsoft/vstest/pull/4497

I have addressed this by explicitly setting the `IsTestProject` property to false in Testcontainers.Commons.csproj in b878d3f359b2f6e6d64ed074dc27833b1d8b74aa. The test run should not be hanging indefinitely anymore.

HofmeisterAn commented:
Here applies the same as for https://github.com/testcontainers/testcontainers-dotnet/pull/908. We do not want to increase coupling in our module tests. I will close the PR for now. If you have any other thoughts on that topic, please consider reopening it again.
